web/api/queries/query.c

```diff
@@ -758,118 +693,162 @@ static inline void rrdr_done(RRDR *r, long rrdr_line) {
+            else
+                common_update_every = MIN(update_every, common_update_every);
+        }
+
+        weight[tier] = query_plan_points_coverage_weight(common_first_t, common_last_t, common_update_every, after_wanted, before_wanted, points_wanted);
     }
 
-    int best_tier = 0;
-    for(int tier = 1; tier < storage_tiers ; tier++) {
+    size_t best_tier = 0;
+    for(size_t tier = 1; tier < storage_tiers ; tier++) {
         if(weight[tier] >= weight[best_tier])
             best_tier = tier;
     }
 
     if(weight[best_tier] == -LONG_MAX)
         best_tier = 0;
 
-    //buffer_sprintf(wb, ": final best tier %d", best_tier);
-    //internal_error(true, "%s", buffer_tostring(wb));
-    //buffer_free(wb);
-
     return best_tier;
 }
 
-static int rrdset_find_natural_update_every_for_timeframe(RRDSET *st, time_t after_wanted, time_t before_wanted, long points_wanted, RRDR_OPTIONS options, int tier) {
-    int ret = st->update_every;
-
-    if(unlikely(!rrdset_number_of_dimensions(st)))
-        return ret;
+static time_t rrdset_find_natural_update_every_for_timeframe(QUERY_TARGET *qt, time_t after_wanted, time_t before_wanted, size_t points_wanted, RRDR_OPTIONS options, size_t tier) {
+    size_t best_tier;
+    if((options & RRDR_OPTION_SELECTED_TIER) && tier < storage_tiers)
+        best_tier = tier;
+    else
+        best_tier = rrddim_find_best_tier_for_timeframe(qt, after_wanted, before_wanted, points_wanted);
 
-    RRDDIM *first_rd = NULL;
-    rrddim_foreach_read(first_rd, st) break; rrddim_foreach_done(first_rd);
-    if(!first_rd)
-        return ret;
+    // find the db minimum update every for this tier for all metrics
+    time_t common_update_every;
+    for(size_t i = 0, used = qt->query.used; i < used ; i++) {
+        QUERY_METRIC *qm = &qt->query.array[i];
 
-    int best_tier;
-    if(options & RRDR_OPTION_SELECTED_TIER && tier >= 0 && tier < storage_tiers)
-        best_tier = tier;
-    else {
-        best_tier = rrddim_find_best_tier_for_timeframe(first_rd, after_wanted, before_wanted, points_wanted);
-    }
+        time_t update_every = qm->tiers[best_tier].db_update_every;
 
-    if(!first_rd->tiers[best_tier]) {
-        internal_error(
-            true,
-            "QUERY: tier %d on chart '%s', is not initialized", best_tier, rrdset_name(st));
-    }
-    else {
-        ret = (int)first_rd->tiers[best_tier]->tier_grouping * (int)st->update_every;
-        if(unlikely(!ret)) {
-            internal_error(
-                true,
-                "QUERY: update_every calculated to be zero on chart '%s', tier_grouping %d, update_every %d",
-                rrdset_name(st), first_rd->tiers[best_tier]->tier_grouping, st->update_every);
-
-            ret = st->update_every;
-        }
+        if(!i)
+            common_update_every = update_every;
+        else
+            common_update_every = MIN(update_every, common_update_every);
     }
 
-    return ret;
+    return common_update_every;
```

Comment:
Compiler is considering that loop cannot be executed and it is reporting:

```sh
web/api/queries/query.c: In function 'rrdset_find_natural_update_every_for_timeframe':
web/api/queries/query.c:851:12: warning: 'common_update_every' may be used uninitialized [-Wmaybe-uninitialized]
  851 |     return common_update_every;
      |            ^~~~~~~~~~~~~~~~~~~
```

Reply:
solved